src/ordered_dict.jl

```diff
@@ -21,25 +25,28 @@ type OrderedDict{K,V} <: Associative{K,V}
     ndel::Int
     dirty::Bool
 
-    function OrderedDict()
-        new(zeros(Int32,16), Array(K,0), Array(V,0), 0, false)
-    end
+    OrderedDict() = new(zeros(Int32,16), Array(K,0), Array(V,0), 0, false)
+
     function OrderedDict(kv)
-        h = OrderedDict{K,V}()
-        for (k,v) in kv
-            h[k] = v
-        end
-        return h
-    end
-    OrderedDict(p::Pair) = setindex!(OrderedDict{K,V}(), p.second, p.first)
-    function OrderedDict(ps::Pair...)
-        h = OrderedDict{K,V}()
-        sizehint!(h, length(ps))
-        for p in ps
-            h[p.first] = p.second
+        (n, slotsz) = get_n_slotsz(kv)
+        h = new(zeros(Int32,slotsz), Array(K,n), Array(V,n), 0, false)
+        if n > 0
+            i = 0
+            for (k,v) in kv
+                i = i + 1
+                index = ht_keyindex2(h, k)
+                @inbounds h.keys[i] = k
+                @inbounds h.vals[i] = v
+                @inbounds h.slots[-index] = i
```

Comment:
Closer, but there needs to be a little more defensive programming here.  Since `kv` is an arbitrary set of key-value pairs, it's possible that a key is repeated, so you'll need to check the return value of `index` and respond accordingly.  You should double check against `setindex!` to make sure there aren't any other corner cases.